content/posts/01-gsoc-2020.rst

```diff
@@ -0,0 +1,198 @@
+========
+
+Before I started working on the HID report descriptor parser, I spent a few days
+with the HID specification. This allowed me to familiarize myself more with the
+intricacies of the of HID and the report descriptor format. It helped a lot
+designing the data structures and the parser. Even with this preparation, I did
+not get it right at first, but I got it close enough that it was easy to fix.
+
+After having the parser implemented, I started working on the data annotations.
+Along the way, I natuarally tweaked the parser to better adapt to our use-case.
+The step of writting the annotations also presented its issues.
+Wireshark operates data at a byte level, but HID operates at a bit level. There
+were already some internal ways to try to get around this issue, but they needed
+to be expanded to properly cater to the needs of HID. Wireshark has encontered
+some protocols needed to dynamically annotate data on a bit level, but their
+requirements were simpler, they mostly wanted to annotate a fixed data type.
+HID, however, can have long data fields, so our main question here was how to
+annotate byte arrays that are bit limited? The same approach used in other types
+of data couldn't be applied here.
+
+Overcoming these issues, I think we got some nice end results.
+
+
+.. figure:: ../../images/posts/01/mouse-dissection.png
+
+    Image 1 - Mouse data dissection
+
+
+.. figure:: ../../images/posts/01/keyboard-dissection.png
+
+    Image 2 - Keyboard data dissection
+
+
+.. figure:: ../../images/posts/01/joystick-dissection.png
+
+    Image 3 - Joystick data dissection
+
+
+.. figure:: ../../images/posts/01/vendor-dissection.png
+
+    Image 4 - Vendor data dissection
+
+
+As you can see, annotations for some fields are still missing. I will keep
+adding more fields as time allows. What is important here is that adding
+support for more fields is easy.
+
+
+Gotchas
+=======
+
+Missing descriptors
+-------------------
+
+To dissect the HID data we need access to the report descriptor, which tells us
+how the reports are structured. This is also the case of for eg. the USB
+endpoint descriptors, we need to look at the endpoint descriptors to understand
+the endpoints. But there is an issue, we are dissecting from a capture, either
+live or from a file, we may not have access to the device to fetch this
+descriptors. Because of this, we need the packets that ask the device for this
+information to be included in the capture.
+Fortunately, libpcap_ asks the device for the USB endpoint descriptors when you
+start recording, but unfortunately, it doesn't do the same for the HID report
+descriptors.
+
+So, if you want to look at HID data dissections, you need to start recording
+before pluging in the device, this way you will also capture the packets where
+the operating system asks the device for the HID descriptors.
+
+USB dissector bug
+-----------------
+
+While working on this project, I found some cases where the data wasn't being
```

Comment:
This looks contradictory. How can data be not dissected and be dissected twice at the same time? The bug is about data being dissected twice - first by the target dissected, and then by the generic handler. This can be confusing to the user, but does not prevent the HID dissector from getting the data.

Reply:
Yes, it does not prevent it from getting the data, but the tvb offset it gets is wrong IIRC.

What if I change it to "I found some cases where the data wasn't being *properly* dissected"?

Reply:
Change seems fine. I think you also found other issues that were fixed, or do I remember incorrectly?

Reply:
Thanks.

Not sure, if I did it was inside the HID dissector and was very minimal. Perhaps the exponent field dissection? I don't recall.

Reply:
The bug that you found and I fixed was https://code.wireshark.org/review/37328

Reply:
Ah! Totally forgot about it :blush:

Do you want me to add it here? It was fairly minimal.

Reply:
I think it makes sense to mention it. If not the GSoC project, it would probably remain unnoticed for many months.

Reply:
Sure.